Datatype upload integration test and fixes

mvdbeek opened the pull request:
Hopefully catches issues like #6849 earlier.

This also tightens up fastqsanger sniffing, introduces a fastqcssanger sniffer (which is unambiguously not fastqsanger), fixes the memepsp sniffer that was fairly broken, fixed a segfault when uploading bcf.uncompressed, fixes datatypes that set binary text or numpy integer/floats on MetadataElement objects on python 3, and skips newline / space2tab conversion when check_binary returns a false negative but the datatype was detected as being binary.

To test additional new datatypes just drop a test file into lib/galaxy/datatypes/test and it will be uploaded and checked. This should catch sniffers with too many false positives.

mvdbeek committed:
Name datatype test files according to datatype extension

mvdbeek committed:
Add integration test for uploading datasets of known datatype

We upload everything in lib/galaxy/datatypes/test and make sure
that we get the correct datatype.
For datatypes that have a sniffer we use `auto`, otherwise we set the
correct datatype.

This is fairly basic for now, but it would be trivial to extend this
to also verify file integrity and metadata elements as well as testing
options that affect uploads, like `auto_decompress` settings.

mvdbeek committed:
Fix uncompressed bcf

This actually also fails, but at least doesn't crash the interpreter.

mvdbeek committed:
Tighten up memepsp sniffer

Values need to be floats, but float("1") also doesn't raise an
exception. This prevents 454qual files to be sniffed as memepsp.

mvdbeek committed:
Move tblastn_four_human_vs_rhodopsin.xml to tblastn_four_human_vs_rhodopsin.blastxml

mvdbeek committed:
Change more biom2 instances

mvdbeek committed:
Rename gff3 test file

mvdbeek committed:
Delete unused test file

mvdbeek committed:
Rename interval1.bed to interv1.bed

Otherwise guess_datatype would expect interval as the datatype.

mvdbeek committed:
Rename 2.shuffled.bam to 2.shuffled.unsorted.bam

mvdbeek committed:
Tighten up fastq sniffers significantly

This requires the sequence regex to match over the entire sequence,
which precludes fastqcssanger files to be sniffed as fastqsanger
(fixes https://github.com/galaxyproject/galaxy/issues/6734).
Further adds a sniffer for FastqCSSanger, which contains
numbers in the sequence and so shouldn't interfere with the other
fastq datatypes.

mvdbeek committed:
Map fastqsolex to fastq datatype in integration test

mvdbeek commented:
Already caught issues in bcf_uncompress, meme_psp and fastqcssanger, and I'm sure validating the set metadata will catch more still.

mvdbeek committed:
Drop index file for bcf_uncompressed

Even `bcftools index` complains that the file is not compressed.
I think this is just not supported.

mvdbeek committed:
Fix totally broken memepsp sniffer

The sniffer could never reach 100 sniffed lines,
and the test file contains an empty line, so the
comment and code that says that empty lines are not
valid is wrong, but was never called.

mvdbeek committed:
Verify state of uploaded datasets and download datasets to verify that the file is intact after upload

mvdbeek committed:
Add a fastqcssanger test file

mvdbeek committed:
Missed one idpDB instance

mvdbeek committed:
Move xls file to excel.xls and change xls to be excel.xls

mvdbeek committed:
Fix test logic

mvdbeek committed:
Fix up various test datasets

mvdbeek committed:
Fix minor typo

nsoranzo commented on lib/galaxy/datatypes/sniff.py:
```diff
@@ -359,7 +359,7 @@ def guess_ext(fname, sniff_order, is_binary=False):
     >>> guess_ext(fname, sniff_order)
     'unsorted.bam'
-    >>> fname = get_test_fname('test.idpDB')
+    >>> fname = get_test_fname('test.idpdb')
```
This is the filename, which was correct before. Revert?

mvdbeek replied:
No, this is intentional, to map the file extension to the datatype extension. I've also moved the file.

mvdbeek committed:
Safe JSON encoding of MetadataParameters, JSONType

This fixes a couple of upload tests on python 3 (loom, Biom2).

mvdbeek committed:
Skip newlines/space2tab conversion for binary datatypes that fail is_binary check

mvdbeek committed:
Fix newline conversion on python 3 if no encoding is set

mvdbeek commented:
@galaxybot test py3

mvdbeek committed:
Fix check_html on python 3

We can open the file in binary mode and then use unicodify on the lines,
as we have to do that anyway if `chunks` is provided.
Also adds the unicode flag to re.compile.

mvdbeek commented:
https://jenkins.galaxyproject.org/job/docker-api-py3/163/testReport/test.api.test_workflows/WorkflowsApiTestCase/test_workflow_run_zip_collections/ is probably unrelated but needs fixing

nsoranzo commented on lib/galaxy/datatypes/binary.py:
```diff
@@ -657,7 +657,8 @@ def set_meta(self, dataset, overwrite=True, **kwd):
         try:
-            pysam.bcftools.index(dataset_symlink)
+            cmd = ['python', '-c', "import pysam.bcftools; pysam.bcftools.index('%s')" % (dataset_symlink)]
+            subprocess.check_call(cmd)
```
This is not needed any more after commit e561267bfc6e36853b4f56ebabda7acbdfa9eeb, right?

mvdbeek replied:
It's still good to do it this way because you may set metadata on an invalid bcf file, causing a segfault. If metadata is being set externally it's not a huge issue, but the internal metadata setting would just kill the interpreter.

mvdbeek commented:
Not sure what's going on with the selenium tests, but at least the failing API test will be fixed by ordering the workflow conversion as done in gxformat2 >= 0.7.0

mvdbeek committed:
Tighten up mothur and phylip sniffers

mvdbeek committed:
Don't sniff simple json types

These should instead be text.

mvdbeek commented:
The selenium python 3 docker tests pass fine on my laptop, I guess that the disk is full on the node that runs the tests. So all of the remaining test failures are false negatives and the flakey python 3 API tests would be fixed by upgrading gxformat2.

jmchilton commented on lib/galaxy/datatypes/test/2.txt:
```diff
@@ -1 +1 @@
-a	2
\ No newline at end of file
+a	2
```
Is this changing the newline used? Why?

mvdbeek replied:
Yes, so that we can verify that a upload-> download cycle still results in the same file (using a hash). We could also allow some diff, or skip that test.

mvdbeek replied:
Or test without newline conversion

mvdbeek commented:
@galaxybot test py3

mvdbeek committed:
Use raw strings in regex and allow lowercase fastq sequences

mvdbeek committed:
Fix UploadProblemException

mvdbeek commented:
@galaxybot test py3

mvdbeek committed:
Avoid exception if no match

This fixes the sniffing unit test, the datatype framework would just
continue anyway.

nsoranzo commented:
Need to rebase again unfortunately.

mvdbeek committed:
Add newline to lib/galaxy/datatypes/test/5e5z.pqr

jmchilton commented:
Wonderful!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

martenson commented:
This is damn great, thank you so much @mvdbeek !
